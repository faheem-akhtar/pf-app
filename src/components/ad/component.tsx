import { FunctionComponent } from 'react';

import styles from './ad.module.scss';

export const AdComponent: FunctionComponent<{ id: string }> = ({ id }) => <div data-ad className={styles.ad} id={id} />;
