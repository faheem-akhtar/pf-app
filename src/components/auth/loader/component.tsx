import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './auth-loader.module.scss';

export const AuthLoaderComponent = ({
  isEnabled,
  isCentered,
}: {
  isEnabled: boolean;
  isCentered: boolean;
}): JSX.Element | null =>
  isEnabled ? (
    <div className={styles.loader1}>
      <div
        className={domClassMerge(styles['loader1__spinner'], isCentered ? styles['loader1__spinner--centered'] : '')}
      >
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--1'])} />
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--2'])} />
        <div className={domClassMerge(styles['loader1__bounce'], styles['loader1__bounce--3'])} />
      </div>
    </div>
  ) : null;
