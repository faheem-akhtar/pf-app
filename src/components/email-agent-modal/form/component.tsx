import { configEmailAgentMustAcceptTerms } from 'config/email-agent/must-accept-terms';
import { configEmailAgentReceiveAdsEnabled } from 'config/email-agent/receive-ads-enabled';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';

import { EmailAgentModalFormAcceptConditionsErrorMessageTemplate } from './accept-conditions-error-message-template';
import { EmailAgentModalFormComponentPropsInterface } from './component-props.interface';
import styles from './email-agent-modal-form.module.scss';
import { FormFieldsEnum } from './fields.enum';
import { EmailAgentModalFormWidgetAcceptTermsComponent } from './widget/accept-terms-component';
import { EmailAgentModalFormWidgetEmailAlertComponent } from './widget/email-alert-component';
import { EmailAgentModalFormWidgetEmailComponent } from './widget/email-component';
import { EmailAgentModalFormWidgetMessageComponent } from './widget/message-component';
import { EmailAgentModalFormWidgetNameComponent } from './widget/name-component';
import { EmailAgentModalFormWidgetPhoneComponent } from './widget/phone-component';
import { EmailAgentModalFormWidgetReCaptchaComponent } from './widget/re-captcha-component';
import { EmailAgentModalFormWidgetReceiveAdvertisingComponent } from './widget/receive-advertising-component';

export const EmailAgentModalFormComponent = ({
  fieldsValue,
  setFieldsValue,
  onSubmit,
  errors,
  loading,
  t,
}: EmailAgentModalFormComponentPropsInterface): JSX.Element => {
  const hasTermsConditionError = !!errors[FormFieldsEnum.acceptTerms] || !!errors[FormFieldsEnum.receiveAdvertising];

  return (
    <form data-testid='email-agent-form' className={styles.form} onSubmit={onSubmit}>
      <EmailAgentModalFormWidgetNameComponent
        t={t}
        value={fieldsValue.name}
        error={errors[FormFieldsEnum.name]}
        onChange={(newValue): void => setFieldsValue({ ...fieldsValue, name: newValue })}
      />
      <EmailAgentModalFormWidgetEmailComponent
        t={t}
        value={fieldsValue.email}
        error={errors[FormFieldsEnum.email]}
        onChange={(newValue): void => setFieldsValue({ ...fieldsValue, email: newValue })}
      />
      <EmailAgentModalFormWidgetPhoneComponent
        t={t}
        value={fieldsValue.phone}
        error={errors[FormFieldsEnum.phone]}
        onChange={(newValue): void => setFieldsValue({ ...fieldsValue, phone: newValue })}
      />
      <EmailAgentModalFormWidgetMessageComponent
        t={t}
        value={fieldsValue.message}
        error={errors[FormFieldsEnum.message]}
        onChange={(newValue): void => setFieldsValue({ ...fieldsValue, message: newValue })}
      />
      <EmailAgentModalFormWidgetReCaptchaComponent />
      <EmailAgentModalFormWidgetEmailAlertComponent
        t={t}
        value={fieldsValue.emailAlert}
        onChange={(newValue): void => setFieldsValue({ ...fieldsValue, emailAlert: newValue })}
      />
      {configEmailAgentMustAcceptTerms && hasTermsConditionError && (
        <EmailAgentModalFormAcceptConditionsErrorMessageTemplate error={errors[FormFieldsEnum.acceptTerms]} />
      )}
      {configEmailAgentMustAcceptTerms && (
        <EmailAgentModalFormWidgetAcceptTermsComponent
          t={t}
          value={fieldsValue.acceptTerms}
          error={errors[FormFieldsEnum.acceptTerms]}
          onChange={(newValue): void => setFieldsValue({ ...fieldsValue, acceptTerms: newValue })}
        />
      )}
      {configEmailAgentReceiveAdsEnabled && (
        <EmailAgentModalFormWidgetReceiveAdvertisingComponent
          t={t}
          value={fieldsValue.receiveAdvertising}
          error={errors[FormFieldsEnum.receiveAdvertising]}
          onChange={(newValue): void => setFieldsValue({ ...fieldsValue, receiveAdvertising: newValue })}
        />
      )}

      <ButtonTemplate
        type='submit'
        size={ButtonSizeEnum.regular}
        componentType={ButtonComponentTypeEnum.primary}
        className={styles.button}
        loading={loading}
      >
        {t('agent-modal/cta-send-message')}
      </ButtonTemplate>
    </form>
  );
};
