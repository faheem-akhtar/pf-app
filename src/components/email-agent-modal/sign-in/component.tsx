import { useContext, useRef } from 'react';

import { UserContext } from 'context/user/context';

import { EmailAgentModalSignInComponentPropsInterface } from './component-props.interface';
import { EmailAgentModalSignInTemplate } from './template';

export const EmailAgentModalSignInComponent = (
  props: EmailAgentModalSignInComponentPropsInterface
): JSX.Element | null => {
  const openAuthRef = useRef<() => void>(() => null);
  const closeAuthRef = useRef<() => void>(() => null);

  const user = useContext(UserContext);

  if (user) return null;

  return <EmailAgentModalSignInTemplate {...props} openAuthRef={openAuthRef} closeAuthRef={closeAuthRef} />;
};
