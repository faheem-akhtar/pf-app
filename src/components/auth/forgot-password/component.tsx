import React, { useState } from 'react';

import { AuthForgotPasswordFieldEnum } from 'components/auth/forgot-password/field.enum';
import { AuthForgotPasswordPropsInterface } from 'components/auth/forgot-password/props.interface';
import { AuthLoaderComponent } from 'components/auth/loader/component';
import { AuthLoginFieldEnum } from 'components/auth/login/field.enum';
import { AuthResetPasswordService } from 'services/auth/reset-password.service';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ErrorMessageComponent } from 'components/error-message/component';
import { formMakeValidator } from 'components/form/make-validator';
import { GoogleRecaptchaService } from 'services/google/recaptcha.service';
import { InputBaseComponent } from 'library/input/base/component';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { useTranslation } from 'helpers/translation/hook';
import { validationEmail } from 'helpers/validation/email';
import { validationRequired } from 'helpers/validation/required';

import styles from '../auth.module.scss';

export const AuthForgotPasswordComponent = (props: AuthForgotPasswordPropsInterface): JSX.Element => {
  const captchaService = GoogleRecaptchaService();
  const { t } = useTranslation();
  const [email, setEmail] = useState('');
  const [errors, setErrors] = useState<Partial<Record<AuthLoginFieldEnum, string>>>({});
  const [errorMessage, setErrorMessage] = useState('');
  const [validators] = useState(() => ({
    [AuthLoginFieldEnum.email]: [
      validationRequired(t('validation/empty-email')),
      validationEmail(t('validation/not-valid-email')),
    ],
  }));
  const [isLoading, setIsLoading] = useState(false);

  /**
   * Validate form fields
   */
  const validate = formMakeValidator(errors, setErrors, validators);

  // TODO-FE[CX-499] - Add success message
  return (
    <div data-testid='AuthForgotPasswordComponent'>
      <div className={styles.heading}>{t('auth/forgot-password')}?</div>
      <AuthLoaderComponent isEnabled={isLoading} isCentered={true} />
      {errorMessage && <ErrorMessageComponent message={errorMessage} />}
      <form
        onSubmit={(e): void => {
          e.preventDefault();

          if (
            !validate({
              [AuthForgotPasswordFieldEnum.email]: email,
            })
          ) {
            setIsLoading(true);
            captchaService.execute().then((captcha_token) => {
              AuthResetPasswordService({
                email,
                captcha_token,
              })
                .then((e) => {
                  if (e.ok) {
                    // Reset server error message
                    setErrorMessage('');
                    // Close modal
                    props.onClose();
                    props.onSuccess();
                  } else {
                    setErrorMessage(e.error?.body || `${t('auth/something-wrong')}! ${t('auth/try-later')}`);
                    captchaService.reset();
                  }
                })
                .finally(() => setIsLoading(false));
            });
          }
          return;
        }}
      >
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('email')}
            type='email'
            value={email}
            error={!!errors[AuthForgotPasswordFieldEnum.email]}
            onChange={(value): void => {
              validate({
                [AuthForgotPasswordFieldEnum.email]: value,
              });
              setEmail(value);
            }}
            errorText={errors[AuthForgotPasswordFieldEnum.email]}
          />
        </div>
        <div className={styles['input-area']}>
          <ButtonTemplate
            className={styles.button}
            type='submit'
            componentType={ButtonComponentTypeEnum.primary}
            size={ButtonSizeEnum.regular}
          >
            {t('auth/reset-password')}
          </ButtonTemplate>
        </div>
        <ReCaptchaComponent />
      </form>
      <div className={domClassMerge(styles.link, styles['create-account'])} onClick={props.onRegister}>
        {t('auth/not-registered-yet')}?&nbsp;{t('auth/create-account')}
      </div>
    </div>
  );
};
