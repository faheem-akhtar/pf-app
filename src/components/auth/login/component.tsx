import React, { useState } from 'react';

import { AuthFacebookService } from 'services/auth/facebook.service';
import { AuthGoogleService } from 'services/auth/google.service';
import { AuthLoaderComponent } from 'components/auth/loader/component';
import { AuthLoginFieldEnum } from 'components/auth/login/field.enum';
import { AuthLoginPropsInterface } from 'components/auth/login/props.interface';
import { AuthLoginService } from 'services/auth/login.service';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ErrorMessageComponent } from 'components/error-message/component';
import { formMakeValidator } from 'components/form/make-validator';
import { GoogleRecaptchaService } from 'services/google-recaptcha/service';
import { IconFacebookTemplate } from 'components/icon/facebook-template';
import { IconGoogleTemplate } from 'components/icon/google-template';
import { InputBaseComponent } from 'library/input/base/component';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { useTranslation } from 'helpers/translation/hook';
import { validationEmail } from 'helpers/validation/email';
import { validationRequired } from 'helpers/validation/required';

import styles from '../auth.module.scss';
const captchaService = new GoogleRecaptchaService();

export const AuthLoginComponent = (props: AuthLoginPropsInterface): JSX.Element => {
  const { t } = useTranslation();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [errors, setErrors] = useState<Partial<Record<AuthLoginFieldEnum, string>>>({});
  const [errorMessage, setErrorMessage] = useState('');
  const [validators] = useState(() => ({
    [AuthLoginFieldEnum.email]: [
      validationRequired(t('validation/empty-email')),
      validationEmail(t('validation/not-valid-email')),
    ],
    [AuthLoginFieldEnum.password]: [validationRequired(t('auth/empty-password'))],
  }));
  const [isLoading, setIsLoading] = useState(false);

  /**
   * Validate form fields
   */
  const validate = formMakeValidator(errors, setErrors, validators);

  return (
    <div>
      <div className={styles.heading}>{t('sign-in')}</div>
      {/* Facebook */}
      <ButtonTemplate
        type='button'
        className={domClassMerge(styles.button, styles.facebook)}
        componentType={ButtonComponentTypeEnum.tertiary}
        size={ButtonSizeEnum.small}
        onClick={(): void => {
          setIsLoading(true);
          AuthFacebookService.signIn()
            .then(props.onClose)
            .catch(() => {
              setErrorMessage(`${t('auth/something-wrong')}! ${t('auth/try-later')}`);
              props.onClose();
            })
            .finally(() => setIsLoading(false));
        }}
        icon={{ component: IconFacebookTemplate, position: ButtonIconPositionEnum.left }}
      >
        {t('auth/sign-in-facebook')}
      </ButtonTemplate>
      {/* Google */}
      <ButtonTemplate
        type='button'
        className={domClassMerge(styles.button, styles.google)}
        componentType={ButtonComponentTypeEnum.tertiary}
        size={ButtonSizeEnum.small}
        onClick={(): void => {
          setIsLoading(true);
          AuthGoogleService.signIn()
            .then(props.onClose)
            .catch(() => {
              setErrorMessage(`${t('auth/something-wrong')}! ${t('auth/try-later')}`);
              props.onClose();
            })
            .finally(() => setIsLoading(false));
        }}
        icon={{ component: IconGoogleTemplate, position: ButtonIconPositionEnum.left }}
      >
        {t('auth/sign-in-google')}
      </ButtonTemplate>
      <div className={styles.divider}>
        <span className={styles.text}>{t('auth/or')}</span>
      </div>
      <AuthLoaderComponent isEnabled={isLoading} isCentered={true} />
      {errorMessage && <ErrorMessageComponent message={errorMessage} />}
      <form
        onSubmit={(e): void => {
          e.preventDefault();

          if (
            !validate({
              [AuthLoginFieldEnum.email]: email,
              [AuthLoginFieldEnum.password]: password,
            })
          ) {
            setIsLoading(true);
            captchaService.execute().then((captcha_token) => {
              AuthLoginService({
                email,
                password,
                captcha_token,
              })
                .then((e) => {
                  if (e.ok) {
                    // Reset server error message
                    setErrorMessage('');
                    // Close modal
                    props.onClose();
                  } else {
                    setErrorMessage(e.error.body || `${t('auth/something-wrong')}! ${t('auth/try-later')}`);
                    captchaService.reset();
                  }
                })
                .finally(() => setIsLoading(false));
            });
          }
          return;
        }}
      >
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('email')}
            type='email'
            value={email}
            error={!!errors[AuthLoginFieldEnum.email]}
            onChange={(value): void => {
              validate({
                [AuthLoginFieldEnum.email]: value,
              });
              setEmail(value);
            }}
            errorText={errors[AuthLoginFieldEnum.email]}
          />
        </div>
        <div className={styles['input-area']}>
          <InputBaseComponent
            placeholder={t('password')}
            type='password'
            value={password}
            error={!!errors[AuthLoginFieldEnum.password]}
            onChange={(value): void => {
              !validate({
                [AuthLoginFieldEnum.password]: value,
              });
              setPassword(value);
            }}
            errorText={errors[AuthLoginFieldEnum.password]}
          />
        </div>
        <div className={styles['input-area']} onClick={props.onForgotPassword}>
          <div className={domClassMerge(styles.link, styles['forgot-password'])}>{t('auth/forgot-password')}?</div>
        </div>
        <div className={styles['input-area']}>
          <ButtonTemplate
            className={styles.button}
            type='submit'
            componentType={ButtonComponentTypeEnum.primary}
            size={ButtonSizeEnum.regular}
          >
            {t('log-in')}
          </ButtonTemplate>
        </div>
        <ReCaptchaComponent />
      </form>
      <div className={domClassMerge(styles.link, styles['create-account'])} onClick={props.onRegister}>
        {t('auth/not-registered-yet')}?&nbsp;{t('auth/create-account')}
      </div>
    </div>
  );
};
