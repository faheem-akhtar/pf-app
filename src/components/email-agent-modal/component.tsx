import { FunctionComponent, useContext, useRef, useState } from 'react';

import { apiEmailAgentFetcher } from 'api/email-agent/fetcher';
import { AuthModalComponent } from 'components/auth/modal/component';
import { formMakeValidator } from 'components/form/make-validator';
import { IconThickSmallCloseTemplate } from 'components/icon/thick/small-close-template';
import { ModalComponent } from 'components/modal/component';
import { propertySerpObfuscatedGetId } from 'components/property/serp/obfuscated/get/id';
import { propertySerpObfuscatedGetName } from 'components/property/serp/obfuscated/get/name';
import { propertySerpObfuscatedGetReference } from 'components/property/serp/obfuscated/get/reference';
import { configEmailAgentEmailAlertCheckedByDefault } from 'config/email-agent/email-alert-checked-by-default';
import { configEmailAgentMustAcceptTerms } from 'config/email-agent/must-accept-terms';
import { configEmailAgentReceiveAdsEnabled } from 'config/email-agent/receive-ads-enabled';
import { UserContext } from 'context/user/context';
import { functionNoop } from 'helpers/function/noop';
import { useTranslation } from 'helpers/translation/hook';
import { validationEmail } from 'helpers/validation/email';
import { validationPhone } from 'helpers/validation/phone';
import { validationRequired } from 'helpers/validation/required';
import { GoogleRecaptchaService } from 'services/google/recaptcha.service';
import { EmailAgentAttributesInterface } from 'types/email-agent/attributes.interface';

import { EmailAgentModalComponentPropsInterface } from './component-props.interface';
import styles from './email-agent-modal.module.scss';
import { EmailAgentModalFormComponent } from './form/component';
import { EmailAgentModalFormComponentPropsInterface } from './form/component-props.interface';
import { EmailAgentModalFormErrorMessageTemplate } from './form/error-message-template';
import { FormFieldsEnum } from './form/fields.enum';
import { FormFieldsValueType } from './form/fields-value.type';
import { EmailAgentModalFormSuccessTemplate } from './form/success-template';
import { EmailAgentModalStatusEnum } from './status.enum';

const captchaService = GoogleRecaptchaService();

export const EmailAgentModalComponent: FunctionComponent<EmailAgentModalComponentPropsInterface> = ({
  openRef,
  property,
}) => {
  const { t } = useTranslation();
  const user = useContext(UserContext);

  const openAuthRef = useRef<() => void>(functionNoop);
  const closeAuthRef = useRef<() => void>(functionNoop);
  const closeRef = useRef<() => void>(functionNoop);

  const getInitialFieldsValue = (): FormFieldsValueType => ({
    name: user ? `${user.first_name} ${user.last_name}` : '',
    email: user ? user.email : '',
    phone: '',
    message: t('agent-modal/default-email-message').replace('#{ref}', propertySerpObfuscatedGetReference(property)),
    emailAlert: configEmailAgentEmailAlertCheckedByDefault,
    acceptTerms: false,
    receiveAdvertising: false,
  });
  const [fieldsValue, setFieldsValue] = useState<FormFieldsValueType>(getInitialFieldsValue());
  const [errors, setErrors] = useState<Partial<Record<FormFieldsEnum, string>>>({});
  const [error, setError] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [validators] = useState(() => {
    const state: Record<string, Array<(value: string) => string>> = {
      [FormFieldsEnum.name]: [validationRequired(t('validation/empty-name'))],
      [FormFieldsEnum.email]: [
        validationRequired(t('validation/empty-email')),
        validationEmail(t('validation/not-valid-email')),
      ],
      [FormFieldsEnum.phone]: [
        validationRequired(t('validation/empty-phone-number')),
        validationPhone(t('validation/not-valid-phone-number')),
      ],
      [FormFieldsEnum.message]: [validationRequired(t('validation/empty-message'))],
    };

    if (configEmailAgentMustAcceptTerms) {
      state[FormFieldsEnum.acceptTerms] = [validationRequired(t('agent-modal/accept-conditions-error-message'))];
    }
    if (configEmailAgentReceiveAdsEnabled) {
      state[FormFieldsEnum.receiveAdvertising] = [validationRequired(t('agent-modal/accept-conditions-error-message'))];
    }
    return state;
  });
  const [status, setStatus] = useState<EmailAgentModalStatusEnum>(EmailAgentModalStatusEnum.opened);
  const validate = formMakeValidator(errors, setErrors, validators);

  const onOpen = (): void => {
    setFieldsValue(getInitialFieldsValue());
    setErrors({});
    setStatus(EmailAgentModalStatusEnum.opened);
  };

  const closeModal = (): void => {
    closeRef.current();
  };

  const onSubmit: EmailAgentModalFormComponentPropsInterface['onSubmit'] = async (e) => {
    e.preventDefault();
    setIsLoading(true);

    const formFields: Record<string, string | boolean> = {
      [FormFieldsEnum.name]: fieldsValue.name,
      [FormFieldsEnum.email]: fieldsValue.email,
      [FormFieldsEnum.phone]: fieldsValue.phone,
      [FormFieldsEnum.message]: fieldsValue.message,
    };

    if (configEmailAgentMustAcceptTerms) {
      formFields[FormFieldsEnum.acceptTerms] = fieldsValue.acceptTerms;
    }
    if (configEmailAgentReceiveAdsEnabled) {
      formFields[FormFieldsEnum.receiveAdvertising] = fieldsValue.receiveAdvertising;
    }

    const hasErrors = validate(formFields);

    if (hasErrors) {
      setIsLoading(false);
      return;
    }

    setStatus(EmailAgentModalStatusEnum.submitting);
    setError('');

    try {
      const captcha_token = await captchaService.execute();

      await apiEmailAgentFetcher({
        propertyId: parseInt(propertySerpObfuscatedGetId(property), 10),
        ...formFields,
        emailAlert: fieldsValue.emailAlert,
        [FormFieldsEnum.captchaToken]: captcha_token,
        autoRegister: !!user,
      } as EmailAgentAttributesInterface);

      setStatus(EmailAgentModalStatusEnum.submitted);
    } catch {
      setError(t('something-wrong-try-again'));
      captchaService.reset();
    } finally {
      setIsLoading(false);
    }
  };

  const body =
    status === EmailAgentModalStatusEnum.submitted ? (
      <EmailAgentModalFormSuccessTemplate closeModal={closeModal} t={t} openAuthRef={openAuthRef} />
    ) : (
      <>
        <h2 className={styles.name}>{propertySerpObfuscatedGetName(property)}</h2>
        <EmailAgentModalFormComponent
          fieldsValue={fieldsValue}
          setFieldsValue={setFieldsValue}
          onSubmit={onSubmit}
          t={t}
          errors={errors}
          loading={isLoading}
        />
      </>
    );

  return (
    <>
      <ModalComponent
        onOpen={onOpen}
        openRef={openRef}
        closeRef={closeRef}
        overlay
        onOverlayClick={(): void => closeRef.current()}
      >
        <div
          data-testid='email-agent-modal-content'
          className={styles.container}
          onClick={(e): void => {
            e.stopPropagation();
          }}
        >
          <header className={styles.header}>
            <h1 className={styles.title}>{t('agent-modal/email-form-title')}</h1>
            <div className={styles.closeButton} onClick={closeModal}>
              <IconThickSmallCloseTemplate class={styles.closeIcon} clipped />
            </div>
          </header>

          {error && <EmailAgentModalFormErrorMessageTemplate error={error} />}
          {body}
        </div>
      </ModalComponent>
      <ModalComponent openRef={openAuthRef} closeRef={closeAuthRef} overlay>
        <AuthModalComponent
          close={(): void => {
            closeAuthRef.current();
          }}
        />
      </ModalComponent>
    </>
  );
};
