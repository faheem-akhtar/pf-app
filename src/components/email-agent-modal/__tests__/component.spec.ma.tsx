/**
 * @jest-environment jsdom
 */

import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { mockModalEnv } from 'mocks/modal-env/mock';
import { mockReactUseSwr } from 'mocks/react/use-swr.mock';
import { countryPhoneCodesStub } from 'stubs/country-phone-codes/stub';
import { propertyStub } from 'stubs/property/stub';

import { PropertySerpObfuscatedType } from 'components/property/serp/obfuscated/type';

import { EmailAgentModalComponent } from '../component';
import { EmailAgentModalComponentPropsInterface } from '../component-props.interface';

describe('EmailAgentModalComponent Marocco', () => {
  const openRef = { current: jest.fn() };
  const property: PropertySerpObfuscatedType = propertyStub();

  let props: EmailAgentModalComponentPropsInterface;

  beforeEach(() => {
    props = {
      property,
      openRef,
    };

    mockModalEnv();
    mockReactUseSwr('en-countries-GET-{"sort":"priority"}', {
      ok: true,
      data: [
        countryPhoneCodesStub({
          code: 'ma',
          name: 'Maroc',
          phoneCode: '+212',
        }),
      ],
    });

    render(<EmailAgentModalComponent {...props} />);
    act(openRef.current);
  });

  it('should render without throwing any errors', () => {
    expect(screen.getByTestId('email-agent-modal-content')).toMatchSnapshot();
  });

  it('should take initial values when modal is opened', () => {
    expect(screen.getByTestId('email-agent-form')).toHaveFormValues({
      emailAlert: false,
      acceptTerms: false,
      receiveAdvertising: false,
    });
  });

  describe('Form Submission', () => {
    it('should display error message for advertising and terms when the checkboxes are not checked', () => {
      userEvent.type(screen.getByLabelText('name'), 'John Doe');
      userEvent.type(screen.getByLabelText('email'), 'john@doe.com');
      userEvent.type(screen.getByPlaceholderText('phone'), '123456789');
      userEvent.type(screen.getByLabelText('message'), 'Message from John Doe');
      userEvent.click(screen.getByRole('button', { name: 'agent-modal/cta-send-message' }));

      const errorText = screen.getByText('agent-modal/accept-conditions-error-message');
      expect(errorText).toBeInTheDocument();

      const acceptTermsCheckbox = screen.getByRole('checkbox', {
        name: 'agent-modal/accept-terms-message-prefix agent-modal/accept-terms-message-link agent-modal/accept-terms-message-suffix',
      });

      userEvent.click(acceptTermsCheckbox);
      expect(acceptTermsCheckbox).toBeChecked();
      userEvent.click(screen.getByRole('button', { name: 'agent-modal/cta-send-message' }));

      expect(errorText).toBeInTheDocument();

      const receiveAdvertisingCheckbox = screen.getByRole('checkbox', {
        name: 'agent-modal/receive-advertising-message',
      });

      userEvent.click(receiveAdvertisingCheckbox);
      expect(acceptTermsCheckbox).toBeChecked();
      userEvent.click(screen.getByRole('button', { name: 'agent-modal/cta-send-message' }));

      expect(errorText).not.toBeInTheDocument();
    });
  });
});
