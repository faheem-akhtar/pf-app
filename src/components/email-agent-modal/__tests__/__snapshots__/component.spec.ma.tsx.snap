// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailAgentModalComponent Marocco should render without throwing any errors 1`] = `
<div
  class="container"
  data-testid="email-agent-modal-content"
>
  <header
    class="header"
  >
    <h1
      class="title"
    >
      agent-modal/email-form-title
    </h1>
    <div
      class="closeButton"
    >
      <svg
        class="closeIcon"
        viewBox="3 3 10 10"
      >
        <path
          clip-rule="evenodd"
          d="M4.34408 3.17475C4.01717 2.92057 3.54448 2.94367 3.24408 3.24408C2.91864 3.56951 2.91864 4.09715 3.24408 4.42259L6.82149 8L3.24408 11.5774C2.91864 11.9028 2.91864 12.4305 3.24408 12.7559C3.54448 13.0563 4.01717 13.0794 4.34408 12.8252L4.42259 12.7559L8 9.17851L11.5774 12.7559L11.6559 12.8252C11.9828 13.0794 12.4555 13.0563 12.7559 12.7559C13.0814 12.4305 13.0814 11.9028 12.7559 11.5774L9.17851 8L12.7559 4.42259C13.0814 4.09715 13.0814 3.56951 12.7559 3.24408C12.4555 2.94367 11.9828 2.92057 11.6559 3.17475L11.5774 3.24408L8 6.82149L4.42259 3.24408L4.34408 3.17475Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </header>
  
  <h2
    class="name"
  >
    Title from agent
  </h2>
  <form
    class="form"
    data-testid="email-agent-form"
  >
    <div
      class="fieldContainer"
    >
      <div
        class="container"
      >
        <div
          class="content"
        >
          <input
            aria-invalid="false"
            aria-label="name"
            class="input"
            id="name"
            name="name"
            type="text"
            value=""
          />
          <div
            class="placeholder"
          >
            <label
              for="name"
            >
              name
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fieldContainer"
    >
      <div
        class="container"
      >
        <div
          class="content"
        >
          <input
            aria-invalid="false"
            aria-label="email"
            class="input"
            id="email"
            name="email"
            type="email"
            value=""
          />
          <div
            class="placeholder"
          >
            <label
              for="email"
            >
              email
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fieldContainer"
    >
      <div
        class="container"
      >
        <div
          class="container"
          role="button"
          tabindex="0"
        >
          <div
            class="body"
          >
            <div
              class="content"
            >
              <div
                class="countryFlag countryFlag--ma"
              />
              <span>
                +212
              </span>
            </div>
            <svg
              class="icon"
              viewBox="4.5 8 16 8"
            >
              <path
                clip-rule="evenodd"
                d="M4.64645 8.14645C4.47288 8.32001 4.4536 8.58944 4.58859 8.78431L4.64645 8.85355L11.6464 15.8536C11.82 16.0271 12.0894 16.0464 12.2843 15.9114L12.3536 15.8536L19.3536 8.85355C19.5488 8.65829 19.5488 8.34171 19.3536 8.14645C19.18 7.97288 18.9106 7.9536 18.7157 8.08859L18.6464 8.14645L12 14.793L5.35355 8.14645C5.17999 7.97288 4.91056 7.9536 4.71569 8.08859L4.64645 8.14645Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <ul
            class="list"
            role="list"
          >
            <li
              class="item"
              role="listitem"
            >
              <div
                class="countryFlag countryFlag--ma"
              />
              Maroc
               
              <span
                class="phoneCode"
              >
                (
                +212
                )
              </span>
            </li>
          </ul>
        </div>
        <div
          class="content"
        >
          <input
            aria-invalid="false"
            aria-label="phone"
            class="input"
            id="phone-number"
            maxlength="12"
            name="phone"
            placeholder="phone"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="fieldContainer"
    >
      <div
        class="container container--textarea"
      >
        <div
          class="content"
        >
          <textarea
            aria-invalid="false"
            aria-label="message"
            class="input input--floating input--textarea"
            id="message"
            name="message"
            type="text"
          >
            agent-modal/default-email-message
          </textarea>
          <div
            class="placeholder placeholder--floating"
          >
            <label
              for="message"
            >
              message
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      id="recaptcha"
    />
    <div
      class="fieldContainer"
    >
      <div
        class="container"
      >
        <input
          aria-labelledby="emailAlert"
          class="input"
          id="email-alert"
          name="emailAlert"
          type="checkbox"
        />
        <label
          class="label"
          for="email-alert"
        >
          agent-modal/email-alert-message
        </label>
      </div>
    </div>
    <div
      class="fieldContainer"
    >
      <div
        class="container"
      >
        <input
          aria-labelledby="acceptTerms"
          class="input"
          id="accept-terms"
          name="acceptTerms"
          type="checkbox"
        />
        <label
          class="label"
          for="accept-terms"
        >
          agent-modal/accept-terms-message-prefix
          <a
            class="link"
            href="terms_conditions-target"
            rel="noreferrer"
            target="_blank"
          >
            agent-modal/accept-terms-message-link
          </a>
          agent-modal/accept-terms-message-suffix
        </label>
      </div>
    </div>
    <div
      class="fieldContainer"
    >
      <div
        class="container"
      >
        <input
          aria-labelledby="receiveAdvertising"
          class="input"
          id="receive-advertising"
          name="receiveAdvertising"
          type="checkbox"
        />
        <label
          class="label"
          for="receive-advertising"
        >
          agent-modal/receive-advertising-message
        </label>
      </div>
    </div>
    <button
      class="button primary regular button"
    >
      agent-modal/cta-send-message
    </button>
  </form>
</div>
`;
